docs/proposals/component-registration.md

```diff
@@ -0,0 +1,294 @@
+
+Since it's seemingly impossible to fix Endpoints Problem \#1 without breaking reverse compatibility, this proposal punts on \#1 (until v2 allows reverse-incompatibility) and focuses on fixing \#2 for now. #13778 handles increasing the scope of Endpoints to cover both Ready and NotReady addresses, which is better than nothing, but not as flexible as having an array of conditions, like other -Status API resources.
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own top-level namespaced API resource (provisionally called **ExternalServiceTarget**). Endpoints Controller will then continue to manage the creation/updates/deletion of Endpoints for internal services based on a selection of Pod instances, but will also do the same for external services based on a selection of ExternalServiceTargets. Endpoints then becomes a (conceptually) read-only API for the user.
+
+
+### API Changes
+
+1. Add new namespaced top level resource: ExternalServiceTarget
+  - Listable, like most other top-level resources
+  - Have both Spec and Status (filterable on both/either)
+  - Spec would have address (ip or host) and ports (optionally named)
+  - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+  - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+  - Similar to Endpoints, but more static. Defines the set of expected service instances, rather than the set of Ready service instances.
+  - Have to be manually added for non-component external Service instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
+  - Analogous to the ServiceSpec.Selector-Pod relationship
+3. Add a new ExternalServiceTarget Controller that handles probing ExternalServiceTargets
+  - Probe results update ExternalServiceTarget.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceTarget Endpoints
+  - Readiness from ExternalServiceTarget.Status.Conditions analogous to PodStatus.Conditions
+5. Modify components to self-register ExternalServiceTargets (one per component instance) instead of an Endpoint
+  - Group ExternalServiceTargets (component instances) with labels (e.g. "type=apiserver", "component=core")
+6. Register component Services (with ServiceSpec.TargetSelector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceTargets.
+7. Modify Service to be support both internal and external targets in the same Service (Pods + ExternalServiceTargets)
```

Comment:
ReferencedName is something I was proposing in #13748 as "a service that is
really something off cluster, represented by a DNS name, the service ports,
and locally represented by the service name in DNS".   However it's
implemented or exposed in the API, the shape is very close to what you're
describing here so I'm in rough agreement.  Agree some of these fields
could crowd out the other fields on service, *but* I could justify there
being a probe on service that is "inclusion" which further constrains the
endpoints selected for the service that could be readiness or liveness or
application defined.

I do think that someone looking at a service directly should be able to
follow the reference of the service off cluster in one step.  For instance,
if I have "oracle-db.foo.mycluster.com" that service "db" is a standin for,
feels like something i should be able to dereference from the service
directly, vs having to hop to a target object.

On Wed, Sep 16, 2015 at 1:58 PM, Joe Beda <notifications@github.com> wrote:

> In docs/proposals/component-registration.md
> <https://github.com/kubernetes/kubernetes/pull/13216#discussion_r39663297>
> :
>
> > +  - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
> > +  - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
> > +  - Similar to Endpoints, but more static. Defines the set of expected service instances, rather than the set of Ready service instances.
> > +  - Have to be manually added for non-component external Service instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
> > +  - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
> > +2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
> > +  - Analogous to the ServiceSpec.Selector-Pod relationship
> > +3. Add a new ExternalServiceTarget Controller that handles probing ExternalServiceTargets
> > +  - Probe results update ExternalServiceTarget.Status.Conditions
> > +4. Modify Endpoints Controller to manage ExternalServiceTarget Endpoints
> > +  - Readiness from ExternalServiceTarget.Status.Conditions analogous to PodStatus.Conditions
> > +5. Modify components to self-register ExternalServiceTargets (one per component instance) instead of an Endpoint
> > +  - Group ExternalServiceTargets (component instances) with labels (e.g. "type=apiserver", "component=core")
> > +6. Register component Services (with ServiceSpec.TargetSelector) as part of cluster deployment
> > +  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceTargets.
> > +7. Modify Service to be support both internal and external targets in the same Service (Pods + ExternalServiceTargets)
>
> I'm not sure what you are suggesting here. Is referencedName new?
>
> I *think* you are in agreement but it isn't clear. This proposal is to
> broaden the service spec so that it can also target external targets. We
> are just representing those external targets as a new object that is a peer
> to pods (ExternalServiceTarget). We considered trying to cram everything
> into ServiceSpec but with probes and all it started getting very unwieldy.
>
> I think the idea of a peer to Pod that is pointing off cluster is an
> elegant solution here. If we could do object interfaces/inheritance with
> REST objects, I'd define a new interface/type called ServiceTarget and
> have both Pod and ExternalServiceTarget implement that interface. But as
> REST doesn't have those semantics it is much more "interface by
> convention/documentation".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13216/files#r39663297>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift